### PR TITLE
Fix race condition in the code update test

### LINKED ISSUE #1147: code_update_test fails
https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=8042&view=logs&j=8f3dc89c-3708-5926-47e7-27120a268dab&t=efdbc34d-c6b6-5dc6-e05c-1751d4c47c20&l=3121

```
38: 2020-05-06 10:24:49.715 | INFO     | __main__:run:87 - Waiting 40.0s for a new primary to be elected...
38: 2020-05-06 10:25:29.752 | INFO     | infra.clients:log_request:147 - [127.210.253.201:56400] GET /nodes/getPrimaryInfo (node 3 (node))
38: 2020-05-06 10:25:29.764 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'primary_host': '127.109.220.177', 'primary_id': 6, 'primary_port': '42250'}, 'error': None, 'commit': 45, 'term': 4, 'global_commit': 45}
38: 2020-05-06 10:25:29.764 | INFO     | __main__:run:91 - Waited, new_primary is 6
38: 2020-05-06 10:25:29.764 | INFO     | __main__:run:93 - Adding another node to the network
38: 2020-05-06 10:25:29.765 | INFO     | infra.clients:log_request:147 - [127.210.253.201:56400] GET /nodes/getPrimaryInfo (node 3 (node))
38: 2020-05-06 10:25:29.776 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'primary_host': '127.109.220.177', 'primary_id': 6, 'primary_port': '42250'}, 'error': None, 'commit': 45, 'term': 4, 'global_commit': 45}
38: 2020-05-06 10:25:29.776 | INFO     | infra.remote:_rc:395 - [127.56.96.88] rm -rf /mnt/build/2/s/build/workspace/code_update_test_7
38: 2020-05-06 10:25:29.780 | INFO     | infra.remote:_rc:395 - [127.56.96.88] mkdir -p /mnt/build/2/s/build/workspace/code_update_test_7
38: 2020-05-06 10:25:29.785 | INFO     | infra.remote:_rc:395 - [127.56.96.88] ln -s /mnt/build/2/s/build/cchost /mnt/build/2/s/build/workspace/code_update_test_7/cchost
38: 2020-05-06 10:25:29.788 | INFO     | infra.remote:_rc:395 - [127.56.96.88] ln -s /mnt/build/2/s/build/liblogging.patched.enclave.so.debuggable /mnt/build/2/s/build/workspace/code_update_test_7/liblogging.patched.enclave.so.debuggable
38: 2020-05-06 10:25:29.790 | INFO     | infra.remote:_rc:395 - [127.56.96.88] cp /mnt/build/2/s/build/workspace/code_update_test_common/networkcert.pem /mnt/build/2/s/build/workspace/code_update_test_7
38: 2020-05-06 10:25:29.793 | INFO     | infra.remote:start:432 - [127.56.96.88] cd /mnt/build/2/s/build/workspace/code_update_test_7 && ./cchost --enclave-file=./liblogging.patched.enclave.so.debuggable --enclave-type=debug --node-address=127.56.96.88:33880 --rpc-address=127.56.96.88:38112 --public-rpc-address=127.56.96.88:38112 --ledger-file=7.ledger --node-cert-file=7.pem --host-log-level=info --raft-election-timeout-ms=20000 --consensus=raft --worker-threads=0 join --network-cert-file=networkcert.pem --target-rpc-address=127.109.220.177:42250 --join-timer=4000 1> /mnt/build/2/s/build/workspace/code_update_test_7/out 2> /mnt/build/2/s/build/workspace/code_update_test_7/err (env: {'OE_LOG_LEVEL': 'WARNING'})
38: 2020-05-06 10:25:30.698 | INFO     | infra.remote:_rc:395 - [127.56.96.88] cp /mnt/build/2/s/build/workspace/code_update_test_7/7.pem /mnt/build/2/s/build/workspace/code_update_test_common/7.pem
38: 2020-05-06 10:25:30.700 | INFO     | infra.node:_start:183 - Remote 7 started
38: 2020-05-06 10:25:30.701 | INFO     | infra.clients:log_request:147 - [127.210.253.201:56400] GET /nodes/getPrimaryInfo (node 3 (node))
38: 2020-05-06 10:25:30.724 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'primary_host': '127.109.220.177', 'primary_id': 6, 'primary_port': '42250'}, 'error': None, 'commit': 45, 'term': 4, 'global_commit': 45}
38: 2020-05-06 10:25:30.724 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /members/read {'table': 'ccf.nodes', 'key': 7} (node 6 (member 0))
38: 2020-05-06 10:25:33.728 | WARNING  | infra.consortium:wait_for_node_to_exist_in_store:408 - Node 7 has not been recorded in the store yet
38: 2020-05-06 10:25:34.229 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /members/read {'table': 'ccf.nodes', 'key': 7} (node 6 (member 2))
38: 2020-05-06 10:25:35.322 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'cert': [45, 45, 45, 45, 45, 66, 69, 71, 73, 78, 32, 67, 69, 82, 84, 73, 70, 73, 67, 65, 84, 69, 45, 45, 45, 45, 45, 10, 77, 73, 73, 66, 121, 106, 67, 67, 65, 85, 54, 103, 65, 119, 73, 66, 65, 103, 73, 82, 65, 79, 67, 98, 109,...
38: 2020-05-06 10:25:35.323 | INFO     | infra.clients:log_request:147 - [127.210.253.201:56400] GET /nodes/getPrimaryInfo (node 3 (node))
38: 2020-05-06 10:25:35.341 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'primary_host': '127.109.220.177', 'primary_id': 6, 'primary_port': '42250'}, 'error': None, 'commit': 47, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.341 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /members/read {'table': 'ccf.nodes', 'key': 7} (node 6 (member 2))
38: 2020-05-06 10:25:35.379 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'cert': [45, 45, 45, 45, 45, 66, 69, 71, 73, 78, 32, 67, 69, 82, 84, 73, 70, 73, 67, 65, 84, 69, 45, 45, 45, 45, 45, 10, 77, 73, 73, 66, 121, 106, 67, 67, 65, 85, 54, 103, 65, 119, 73, 66, 65, 103, 73, 82, 65, 79, 67, 98, 109,...
38: 2020-05-06 10:25:35.379 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /members/propose {'parameter': 7, 'script': {'text': '\n        tables, node_id = ...\n        return Calls:call("trust_node", node_id)\n        '}, 'ballot': {'text': 'return true'}} (node 6 (member 1)) [signed]
38: 2020-05-06 10:25:35.405 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'proposal_id': 10, 'proposer_id': 1, 'state': 'OPEN'}, 'error': None, 'commit': 48, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.405 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /members/vote {'ballot': {'text': '\n        tables, changes = ...\n        return true\n        '}, 'id': 10} (node 6 (member 0)) [signed]
38: 2020-05-06 10:25:35.440 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'proposal_id': 10, 'proposer_id': 1, 'state': 'ACCEPTED'}, 'error': None, 'commit': 49, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.440 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] POST /nodes/mkSign (node 6 (node))
38: 2020-05-06 10:25:35.455 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': True, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.455 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:35.457 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.557 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:35.560 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:35.660 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:35.662 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}

38: 2020-05-06 10:25:37.237 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.337 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.350 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.450 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.458 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.559 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.566 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.667 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.681 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.781 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.790 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:37.890 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:37.906 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.006 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:38.016 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.116 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:38.119 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.219 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:38.234 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.334 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:38.341 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.442 | INFO     | infra.clients:log_request:147 - [127.109.220.177:42250] GET /nodes/tx {'view': 4, 'seqno': 49} (node 6 (node))
38: 2020-05-06 10:25:38.450 | DEBUG    | infra.clients:log_response:160 - {'status': 200, 'result': {'status': 'PENDING'}, 'error': None, 'commit': 50, 'term': 4, 'global_commit': 47}
38: 2020-05-06 10:25:38.550 | ERROR    | infra.ccf:create_and_trust_node:355 - New trusted node 7 failed to join the network
38: 2020-05-06 10:25:38.550 | INFO     | infra.remote:stop:455 - [127.56.96.88] closing
38: 2020-05-06 10:25:38.614 | INFO     | infra.remote:stop:455 - [127.210.253.201] closing
38: 2020-05-06 10:25:38.677 | INFO     | infra.remote:stop:455 - [127.84.4.156] closing
38: 2020-05-06 10:25:38.741 | INFO     | infra.remote:stop:455 - [127.89.88.47] closing
38: 2020-05-06 10:25:38.805 | INFO     | infra.remote:stop:455 - [127.109.220.177] closing
38: 2020-05-06 10:25:38.870 | INFO     | infra.ccf:stop_all_nodes:294 - All remotes stopped...

```

### PR DESCRIPTION
**Pull request by achamayou**
fix #1147 

The gist of the problem is that when a node joins a network, and become a tie-breaker for the consensus (eg. an 8th node joining a 7 node network where 3 nodes are dead, in the case of the code update), there is a race between the join rpc poll and the waiting for the vote to be globally committed.

The new dcap client, which results in faster quote verification seems to have triggered this by making the initial delay shorter in the join RPC, which leads to the vote happening sooner, and the timeout sooner too, before the join RPC has time to poll again.

The code update test is a useful corner case test but not an accurate representation of what we expect a vanilla code update to look like. We ought to refactor it in a similar way to the reconfiguration test, modify it to represent the normal case, and establish variants for pathological cases like this one (which are still interesting!).

### CODE REVIEW

File: tests/infra/member.py [hunk 1]
```diff
@@ -111,7 +111,11 @@ def vote(
             and wait_for_global_commit
         ):
             with remote_node.node_client() as mc:
-                infra.checker.wait_for_global_commit(mc, r.seqno, r.view, True)
+                # If we vote in a new node, which becomes part of quorum, the transaction
+                # can only commit after it has successfully joined and caught up.
+                # Given that the retry timer on join RPC is 4 seconds, anything less is very
+                # likely to time out!
+                infra.checker.wait_for_global_commit(mc, r.seqno, r.view, True, timeout=6)
```
Comment (jumaffre) [inline tests/infra/member.py:118]: Since this path is used for every proposal, would it be more precise to only set this timeout for the `trust_node` proposal?

**Comment by eddyashton** [comment]
There's the following comment on the `--join-timer` arg in `e2e_args.py`:

```
default=4000,  # Set higher than cchost default to avoid swamping joinee with requests during slow quote verification
```

Do we trust the new DCAP client to consistently verify faster? If so, perhaps this default can be removed (the default in `host/main.cpp` is 1000ms, which would 'fix' this without increasing the timeout for all votes).

**Comment by achamayou** [comment]
@eddyashton that's probably one of the things we ought to do, but we have less than 24 hours of testing with the new dcap client, so I suggest we wait a bit first.

Perhaps the first end to end test of our CI should nuke the cache, do a cold and then a warm request, and publish both timings in cimetrics. That seems like it should be easy to add, and we could get a more quantitative view of the responsiveness of quote verification that our current general feeling.

File: tests/infra/member.py (hunk 1 above)
Comment (achamayou) [reply tests/infra/member.py]: Yes, but it also seems like additional complexity. I think the comment explains the intent and will hopefully let us remove this altogether once join is fast enough, and in the meantime, it's just a slightly longer timeout for votes which shouldn't fail that often if at all in the CI (there's no penalty when they pass).
